shell/platform/embedder/embedder_surface_metal_impeller.mm

```diff
@@ -0,0 +1,100 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include <memory>
+#include <utility>
+
+#include "flutter/shell/platform/embedder/embedder_surface_metal_impeller.h"
+
+#include "flutter/fml/logging.h"
+#include "flutter/fml/synchronization/sync_switch.h"
+#include "flutter/shell/gpu/gpu_surface_metal_delegate.h"
+#include "flutter/shell/gpu/gpu_surface_metal_impeller.h"
+#import "flutter/shell/platform/darwin/graphics/FlutterDarwinContextMetalImpeller.h"
+#include "impeller/entity/mtl/entity_shaders.h"
+#include "impeller/entity/mtl/framebuffer_blend_shaders.h"
+#include "impeller/entity/mtl/modern_shaders.h"
+#include "impeller/renderer/backend/metal/context_mtl.h"
+#include "impeller/scene/shaders/mtl/scene_shaders.h"
+
+FLUTTER_ASSERT_NOT_ARC
+
+namespace flutter {
+
+EmbedderSurfaceMetalImpeller::EmbedderSurfaceMetalImpeller(
+    GPUMTLDeviceHandle device,
+    GPUMTLCommandQueueHandle command_queue,
+    MetalDispatchTable metal_dispatch_table,
+    std::shared_ptr<EmbedderExternalViewEmbedder> external_view_embedder)
+    : GPUSurfaceMetalDelegate(MTLRenderTargetType::kMTLTexture),
+      metal_dispatch_table_(std::move(metal_dispatch_table)),
+      external_view_embedder_(std::move(external_view_embedder)),
+      concurrent_loop_(fml::ConcurrentMessageLoop::Create()) {
+  std::vector<std::shared_ptr<fml::Mapping>> shader_mappings = {
+      std::make_shared<fml::NonOwnedMapping>(impeller_entity_shaders_data,
+                                             impeller_entity_shaders_length),
+      std::make_shared<fml::NonOwnedMapping>(impeller_scene_shaders_data,
+                                             impeller_scene_shaders_length),
+      std::make_shared<fml::NonOwnedMapping>(impeller_modern_shaders_data,
+                                             impeller_modern_shaders_length),
+      std::make_shared<fml::NonOwnedMapping>(impeller_framebuffer_blend_shaders_data,
+                                             impeller_framebuffer_blend_shaders_length),
+  };
+  context_ = impeller::ContextMTL::Create(
+      (id<MTLDevice>)device,                     // device
+      (id<MTLCommandQueue>)command_queue,        // command_queue
+      shader_mappings,                           // shader_libraries_data
+      concurrent_loop_->GetTaskRunner(),         // worker_task_runner
+      std::make_shared<fml::SyncSwitch>(false),  // is_gpu_disabled_sync_switch
+      "Impeller Library"                         // library_label
+  );
+
+  valid_ = !!context_;
+}
+
+EmbedderSurfaceMetalImpeller::~EmbedderSurfaceMetalImpeller() = default;
+
+bool EmbedderSurfaceMetalImpeller::IsValid() const {
+  return valid_;
+}
+
+std::unique_ptr<Surface> EmbedderSurfaceMetalImpeller::CreateGPUSurface() API_AVAILABLE(ios(13.0)) {
+  if (@available(iOS 13.0, *)) {
```

Comment:
I don't see the point of this availability guard. There is no call here that the compiler will complain about. And we support targets below 13.0.

Reply:
Whoops, this is because I just duplicated the Skia variation and modified relevant parts. Removed.

Reply:
It turns out we need the check because CAMetalLayer is iOS 13+

Reply:
The docs say 8.0. Perhaps its the iOS simulator?
<img width="153" alt="Screenshot 2023-05-31 at 10 38 34 AM" src="https://github.com/flutter/engine/assets/44085/6b95bdd7-4db1-47aa-8363-0c9116173439">

Reply:
We should copy this sort of macro into formats_mtl and use it for the availability check. https://github.com/google/skia/blob/a8c3132457176d453c063247eaa6e2cab1cb700e/include/gpu/mtl/GrMtlTypes.h#L28

Reply:
So:

```
#if TARGET_OS_SIMULATOR
#define IMPELLER_METAL_AVAILABLE @available(macos(10.11), ios(13.0))
#else  // TARGET_OS_SIMULATOR
#define IMPELLER_METAL_AVAILABLE @available(macos(10.11), ios(8.0))
#endif  // TARGET_OS_SIMULATOR
```

Have macros for `IMPELLER_MACOS_METAL_AVAILABLE`, `IMPELLER_IOS_METAL_AVAILABLE` set to the appropriate versions.